src/Dependencies/Collections/SegmentedList`1.cs

```diff
@@ -502,7 +529,17 @@ internal void Grow(int capacity)
             // If the computed capacity is still less than specified, set to the original argument.
             // Capacities exceeding Array.MaxLength will be surfaced as OutOfMemoryException by Array.Resize.
             if (newCapacity < capacity)
+            {
                 newCapacity = capacity;
+            }
+            else
+            {
+                var segmentSize = SegmentedArrayHelper.GetSegmentSize<T>();
+
+                // If caller didn't request a large capacity increase, limit the increase to a single page
+                if (newCapacity > segmentSize)
+                    newCapacity = (((capacity - 1) / segmentSize) + 1) * segmentSize;
```

Comment:
Let's move this entire section, plus any changes to the logic, to instead be part of the initial calculation of `newCapacity`.

Reply:
➡️ See https://github.com/dotnet/roslyn/pull/75661#discussion_r1824714647